drape_frontend/stylist.cpp

```diff
@@ -430,10 +429,51 @@ bool InitStylist(FeatureType & f, int8_t deviceLang, int const zoomLevel, bool b
   Aggregator aggregator(f, geomType, zoomLevel, keys.size());
   aggregator.AggregateKeys(keys);
 
-  CaptionDescription & descr = s.GetCaptionDescriptionImpl();
-  descr.Init(f, deviceLang, zoomLevel, geomType, aggregator.m_mainTextType, aggregator.m_auxCaptionFound);
+  s.m_captionDescriptor.Init(f, deviceLang, zoomLevel, geomType, aggregator.m_mainTextType, aggregator.m_auxCaptionFound);
+
+  if (aggregator.m_captionRule.m_rule != nullptr)
+  {
+    if (!s.m_captionDescriptor.IsHouseNumberInMainText())
+      aggregator.m_rules.push_back(aggregator.m_captionRule);
+    else
+    {
+      bool isGood = true;
+      if (zoomLevel < scales::GetUpperStyleScale())
+      {
+        if (geomType == feature::GeomType::Area)
+        {
+          // Don't display housenumbers when an area (e.g. a building) is too small.
+          m2::RectD const r = f.GetLimitRect(zoomLevel);
+          isGood = std::max(r.SizeX(), r.SizeY()) > scales::GetEpsilonForHousenumbers(zoomLevel);
+        }
+        else
+        {
+          // Limit point housenumbers display to detailed zooms only (z18-).
+          ASSERT_EQUAL(geomType, feature::GeomType::Point, ());
+          isGood = zoomLevel >= scales::GetAddNewPlaceScale();
+        }
+      }
+
+      if (isGood)
+      {
+        // Use building-address' caption drule to display house numbers.
+        static auto const addressType = cl.GetTypeByPath({"building", "address"});
+        drule::KeysT addressKeys;
+        cl.GetObject(addressType)->GetSuitable(zoomLevel, geomType, addressKeys);
+        if (!addressKeys.empty())
+        {
+          // A caption drule exists for this zoom level.
+          ASSERT(addressKeys.size() == 1 && addressKeys[0].m_type == drule::caption,
+                 ("building-address should contain a caption drule only"));
+          drule::BaseRule const * const dRule = drule::rules().Find(addressKeys[0]);
+          ASSERT(dRule != nullptr, ());
+          aggregator.m_rules.push_back({ dRule, static_cast<float>(addressKeys[0].m_priority), false });
+        }
```

Comment:
I think, we can cache this dRule once somewhere around Classificator or RulesHolder.

Reply:
this drule is zoom-dependent, so will have to cache an array of drules at least

Actually I'm not even 100% sure that `static auto const addressType =` is a good idea, are these types ids guaranteed to be the same across styles? in my test the id was the same for the main and vehicle style.
Looking at the code they should stay the same although classificator is duplicated for each style.

Reply:
@vng with https://github.com/organicmaps/organicmaps/pull/6015 no drules caching is needed.

Reply:
When Apple and Google show house numbers?

Most users use search to find numbers. Less clutter is better in most cases. Are you sure about z17 for house numbers and especially for stand-alone address nodes?

Reply:
I think its more a question to @map-per 

My preference is too make standalone housenumbers visibility low like z18-.

Reply:
You can store building-address object: Classificator::GetAddressObject().
Like it already has GetCoastType or GetStubType/

Reply:
Stylist calls GetObject() for each type in the feature (sometimes 2 times even), so I think a better approach would be to change GetObject() to a map lookup (type -> ClassifObject). WDYT? Making an exception just for the address type seems wrong.

Reply:
The point is not in GetObject itself (can make it better later), but in
```
static auto const addressType = cl.GetTypeByPath({"building", "address"});
```

Reply:
Then I don't understand.. How is e.g. `cl.GetAddressType()` would be better than a static definition you've quoted? This static declaration will be processed just once, right?

Reply:
Right, but I'd prefer to avoid static vars pattern because of the _possible_ (I can't say for sure here) compiler-generated lock guard here.